docs/RationaleRTL-SV.md

```diff
@@ -105,6 +105,114 @@ declarations in limited ways:
  - Interface signals are allowed to be zero bits wide.  They are dropped from
    Verilog emission.
 
+### Endianness: operand ordering and internal representation
+
+Certain operations require ordering to be defined (i.e. `rtl.concat`,
+`rtl.array_concat`, and `rtl.array_create`). There are two places where this
+is relevant: in the MLIR assembly and in the MLIR C++ model.
+
+In MLIR assembly, operands are always listed MSB to LSB (big endian style):
+
+```mlir
+%msb = rtl.constant (0xEF : i8) : i8
+%mid = rtl.constant (0x7 : i4) : i4
+%lsb = rtl.constant (0xA018 : i16) : i16
+%result = rtl.concat %msb, %mid, %lsb : (i8, i4, i16) -> i28
+// %result is 0xEF7A018
+```
+
+**Note**: Integers are always big endian in both bit and byte ordering.
+Operand ordering for `rtl.concat` was chosen to be consistent with simply
+abutting them in lexical order.
+
+```mlir
+%1 = rtl.constant (0x1 : i4) : i4
+%2 = rtl.constant (0x2 : i4) : i4
+%3 = rtl.constant (0x3 : i4) : i4
+%arr123 = rtl.array_create %1, %2, %3 : i4
+// %arr123[0] = 0x3
+// %arr123[1] = 0x2
+// %arr123[2] = 0x1
+
+%arr456 = ... // {0x4, 0x5, 0x6}
+%arr78  = ... // {0x7, 0x8}
+%arr = rtl.array_concat %arr123, %arr456, %arr78 : !rtl.array<3 x i4>, !rtl.array<3 x i4>, !rtl.array<2 x i4>
+// %arr[0] = 0x6
+// %arr[1] = 0x5
+// %arr[2] = 0x4
+// %arr[3] = 0x3
+// %arr[4] = 0x2
+// %arr[5] = 0x1
+```
+
+**Note**: This ordering scheme is unintuitive for anyone expecting C
+array-like ordering. In a C, arrays are laid out with index 0 as the least
+significant value. In the CIRCT _model_ (assembly and C++), it is the
+opposite -- the most significant value is located at the zero index. The
+produced hardware, however, _is_ in C-style order as operands are printed (in
```

Comment:
@lattner Is this correct? It's how it exists now, and I just made this assumption based on the way `ConcatOp` works.

Reply:
And I _think_ it falls out of the rest of the text, but I'd like someone to check me on that.

Reply:
No, this isn't quite right.  This piece:

"In the CIRCT _model_ (assembly and C++), it is the
opposite -- the most significant value is located at the zero index."

Should be something like:

"In the CIRCT _model_ (assembly and C++), it is the
opposite -- the most significant value is on the left (e.g. the first operand is the most significant).  However, the indexing semantics at runtime are that element zero is the least significant (which is lexically on the right".

Reply:
by "zero index" I meant in the operand list. I'll use your wording.

Reply:
done